src/tests/Microsoft.PowerFx.Interpreter.Tests/ThreadingTest.cs

```diff
@@ -0,0 +1,214 @@
+// Licensed under the MIT license.
+
+using System;
+using System.Collections.Generic;
+using System.Reflection;
+using System.Text;
+using Microsoft.PowerFx.Core;
+using Microsoft.PowerFx.Core.Public;
+using Microsoft.PowerFx.Core.Utils;
+using Xunit;
+
+namespace Microsoft.PowerFx.Tests
+{
+    // Do static analysis to look for potential threading issues. 
+    public class ThreadingTest
+    {
+        [Fact]
+        public void CheckIntepreter()
+        {
+            var asm = typeof(RecalcEngine).Assembly;            
+            CheckStatics(asm);
+        }
+
+        [Fact]
+        public void CheckFxCore()
+        {
+            var asm = typeof(Core.Syntax.Nodes.TexlNode).Assembly;
+            CheckStatics(asm);
+        }
+
+        // Verify there are no "unsafe" static fields that could be threading issues.
+        private void CheckStatics(Assembly asm)
+        {
+            var errors = new List<string>();
+
+            var total = 0;
+            foreach (var type in asm.GetTypes())
+            {
+                // Reflecting over fields will also find compiler-generated "backing fields" from static properties. 
+                foreach (var field in type.GetFields(BindingFlags.Static | BindingFlags.Public | BindingFlags.NonPublic))
+                {
+                    var name = $"{type.Name}.{field.Name}";
+                    total++;
+
+                    if (field.Attributes.HasFlag(FieldAttributes.Literal))
+                    {
+                        continue;  // 'const' keyword, 
+                    }
+
+                    if (type.Name.Contains("<>") ||
+                        type.Name.Contains("<PrivateImplementationDetails>"))
+                    {
+                        continue; // exclude compiler generated closures. 
+                    }
+
+                    // Whitelist - these are reviewed and ok
+                    var whitelist = new HashSet<string>
+                    {                     
+                        "CallNode._uniqueInvocationIdNext"
+                    };
+                    if (whitelist.Contains(name))
+                    {
+                        continue;
+                    }
+
+                    // TODO - Drive this list to zero. 
+                    var bugs = new HashSet<Type>
+                    {
+                        typeof(Dictionary<Core.IR.Nodes.UnaryOpKind, Functions.Library.FunctionPtr>),
+                        typeof(Functions.Library.FunctionPtr),
+
+                         // make these fields readonly
+                        typeof(Core.Localization.TexlStrings.StringGetter),
+                        typeof(Core.Localization.ErrorResourceKey),
+                    };
+                    if (bugs.Contains(field.FieldType))
+                    {
+                        continue;
+                    }
+
+                    // TODO - Drive this list to zero. 
+                    var bugNames = new HashSet<string>
```

Comment:
>bugNames [](http://example.com/codeflow?start=24&length=8)

These 2 bug lists give us a denominator. 
Most of these look like just collections / flags to mark readonly; 

There's actually a few interesting cases to review in here...